### PR TITLE
#2403 improvement: EXT MS TLS 1.2

### PR DESCRIPTION
**Pull request by gpotter2**
- rebase of https://github.com/secdev/scapy/pull/2403
- adds a test provided by https://github.com/secdev/scapy/issues/2784
- fixes part of https://github.com/secdev/scapy/issues/2784
- fix https://github.com/secdev/scapy/pull/2403

### CODE REVIEW

File: test/tls.uts [hunk 1]
```diff
@@ -979,6 +979,41 @@ assert(rec_fin.mac == b'\xecguD\xa8\x87$<7+\n\x94\x1e9\x96\xfa')
...
+# Load key
+kf = get_temp_file()
+with open(kf, "wb") as fd:
+    fd.write(key)
+
+ssl_key = PrivKey(kf)
```
Comment (strayge) [inline test/tls.uts:1000]: why needed temporary file?

`PrivKey` can handle bytes as first param.

File: scapy/layers/tls/handshake.py [hunk 2]
```diff
@@ -1286,6 +1305,18 @@ def build(self, *args, **kargs):
...
         return _TLSHandshake.build(self, *args, **kargs)
 
+    def tls_session_update(self, msg_str):
+        """
+        Finalize the EXTMS messages and compute the hash
+        """
+        super(TLSClientKeyExchange, self).tls_session_update(msg_str)
+
+        if self.tls_session.extms:
+            hash_object = self.tls_session.pwcs.hash
```
Comment (strayge) [inline scapy/layers/tls/handshake.py:1315]: Is it correctly fallback to `MD5+SHA1` for `TLS 1.0` - `TLS 1.1` (as said chapter 3 of RFC 7627)?

I've got traffic examples for `Extended-Master-Secret` for TLS 1.0 & TLS 1.1 only with `Encrypt-then-MAC`.  
So can't check it right now.

File: test/tls.uts (hunk 1 above)
Comment (gpotter2) [reply test/tls.uts:1000]: Ha, i didn't even know this

File: scapy/layers/tls/handshake.py (hunk 2 above)
Comment (gpotter2) [reply scapy/layers/tls/handshake.py:1315]: This should be alright, it's the same hash function as the one defined in PRF over here https://github.com/secdev/scapy/blob/54da22f91ce5fd45d30efd0359305f4d1fa448a9/scapy/layers/tls/crypto/prf.py#L178

Comment (gpotter2) [reply scapy/layers/tls/handshake.py:1315]: I'm not a 100% sure, will check that tonight. Should be defined over https://github.com/secdev/scapy/blob/54da22f91ce5fd45d30efd0359305f4d1fa448a9/scapy/layers/tls/session.py#L36 but this is usually pretty alright. This fallback is implemented for some parts like the PRF

Comment (strayge) [reply scapy/layers/tls/handshake.py:1315]: Checked. Current implementation works only for `TLS 1.2`.  
Here fix: https://github.com/strayge/scapy/commit/b6a97dccb880c0285c893a644fc3e57a74f89192

Also for testing I wrote POC of `Encrypt-Then-MAC`: https://github.com/strayge/scapy/commit/aee5f5712f287256a94ea34a46f577b9434374ae

with both this commits successfully decrypted: `TLS 1.0` - `TLS 1.2` with `EMS` + `ETM` extensions (checked on `AES-CBC` & `AES-GCM` suites)

Comment (gpotter2) [reply scapy/layers/tls/handshake.py:1315]: Thanks a lot for the PoC ! Looking good.
There are a few things that require changes (I'll go over that tommorow), such as the case where the server doesn't support ExtMS or EncryptThenMac, despite the client request, but it's a solid start.

Comment (strayge) [reply scapy/layers/tls/handshake.py:1315]: Extensions detection placed in `TLSServerHello` class, so we detect it only if server support it (and client requested it).

Also in PoC missed implementation of `Encrypt-Then-MAC` for stream ciphers (for AEAD ciphers server should not send this extension, according to RFC).

**Comment by p-l-** [comment]
@gpotter2 can you rebase against current master, since #2792 has been merged?

**Comment by strayge** [comment]
@gpotter2 should I made separated PR for Encrypt-then-MAC?

**Comment by gpotter2** [comment]
That would probably be a good thing, thanks for the suggestion !

**Comment by strayge** [comment]
Also I found that current implementation does not work with TLS 1.2 + RC4.
Don't know why.
TLS 1.0 and TLS 1.1 works well with RC4 + extended_master_secret.

openssl 1.1.1g for client and server.
First `Finished` in 3rd packet is not decrypted.
Here is example. Key is the same.
[tls12_rc4_ems.zip](https://github.com/secdev/scapy/files/5211505/tls12_rc4_ems.zip)

File: scapy/layers/tls/handshake.py [hunk 3]
```diff
@@ -1296,6 +1305,26 @@ def build(self, *args, **kargs):
...
+
+        if self.tls_session.extms:
+            to_hash = b''.join(self.tls_session.handshake_messages)
+            if self.tls_session.tls_version >= 0x303:
+                # TLS 1.2 uses the default hash from cipher suite
+                hash_object = self.tls_session.pwcs.hash
+                self.tls_session.session_hash = hash_object.digest(to_hash)
+            else:
+                # Previous TLS version use concatenation of MD5 & SHA1
+                from scapy.layers.tls.crypto.hash import Hash_MD5, Hash_SHA1
```
Comment (strayge) [inline scapy/layers/tls/handshake.py:1322]: It's should be `Hash_SHA` instead of `Hash_SHA1`

File: test/tls.uts [hunk 4]
```diff
@@ -979,6 +979,37 @@ assert(rec_fin.mac == b'\xecguD\xa8\x87$<7+\n\x94\x1e9\x96\xfa')
...
+# Load key
+ssl_key = PrivKey(key)
+
+# Load TLS session
+r1 = TLS(chello_extms)
+r1.tls_session.server_rsa_key = ssl_key
+r2 = TLS(shello_extms, tls_session=r1.tls_session.mirror())
+r3 = TLS(finished_extms, tls_session=r2.tls_session.mirror())
+
+r3
```
Comment (strayge) [inline test/tls.uts:1004]: probably left after debugging

Comment (gpotter2) [reply test/tls.uts:1004]: We've had TLS packets crash while being displayed (here) so I'm going to leave this to trigger that.

**Comment by gpotter2** [comment]
The RC4 example your provided has `encrypt_then_mac` so that's probably the issue.

**Comment by strayge** [comment]
> The RC4 example your provided has `encrypt_then_mac` so that's probably the issue.

`ServerHello` does not contains `encrypt_then_mac`, so it should not be used.

image

**Comment by gpotter2** [comment]
I've provided a fix, feel free to try it out.

**Comment by strayge** [comment]
Thanks! Now it works for all my test cases.

**Pull request merged**